RenderStateNotation/src/RenderStateNotationLoaderImpl.cpp

```diff
@@ -76,51 +78,73 @@ void RenderStateNotationLoaderImpl::LoadPipelineState(const LoadPipelineStateInf
                     LOG_ERROR_AND_THROW("Failed to load shader '", Name, "' for pipeline '", LoadInfo.Name, "'.");
 
-                PipelineShaders.push_back(pShader);
-                return pShader;
+                if (IsAddToCache)
+                    m_ShaderCache.emplace(HashMapStringKey{pShader->GetDesc().Name, false}, pShader);
+
+                return pShader.Detach();
             };
 
             auto FindRenderPass = [&](const char* Name) -> IRenderPass* //
             {
                 if (Name == nullptr)
                     return nullptr;
 
-                VERIFY_EXPR(!pRenderPass);
-                LoadRenderPass(LoadRenderPassInfo{Name, LoadInfo.AddToCache}, &pRenderPass);
+                bool IsAddToCache = LoadInfo.AddToCache;
+                auto Callback     = MakeCallback([&](RenderPassDesc& RenderPassCI) {
+                    if (LoadInfo.ModifyRenderPass != nullptr)
+                        LoadInfo.ModifyRenderPass(RenderPassCI, IsAddToCache, LoadInfo.pModifyRenderPassData);
+                });
+
+                RefCntAutoPtr<IRenderPass> pRenderPass;
+                LoadRenderPass({Name, false, Callback, Callback}, &pRenderPass);
 
                 if (!pRenderPass)
                     LOG_ERROR_AND_THROW("Failed to load render pass '", Name, "' for pipeline '", LoadInfo.Name, "'.");
-                return pRenderPass;
+
+                if (IsAddToCache)
+                    m_RenderPassCache.emplace(HashMapStringKey{pRenderPass->GetDesc().Name, false}, pRenderPass);
+
+                return pRenderPass.Detach();
             };
 
             auto FindResourceSignature = [&](const char* Name) -> IPipelineResourceSignature* //
             {
                 if (Name == nullptr)
                     return nullptr;
 
+                bool IsAddToCache = LoadInfo.AddToCache;
+                auto Callback     = MakeCallback([&](PipelineResourceSignatureDesc& ResourceSignatureCI) {
+                    if (LoadInfo.ModifyResourceSignature != nullptr)
+                        LoadInfo.ModifyResourceSignature(ResourceSignatureCI, IsAddToCache, LoadInfo.pModifyResourceSignatureData);
+                });
+
                 RefCntAutoPtr<IPipelineResourceSignature> pResourceSignature;
-                LoadResourceSignature(LoadResourceSignatureInfo{Name, LoadInfo.AddToCache}, &pResourceSignature);
+                LoadResourceSignature({Name, false, Callback, Callback}, &pResourceSignature);
 
                 if (!pResourceSignature)
                     LOG_ERROR_AND_THROW("Failed to load resource signature '", Name, "' for pipeline '", LoadInfo.Name, "'.");
 
-                PipelineSignatures.push_back(pResourceSignature);
-                return pResourceSignature;
+                if (IsAddToCache)
+                    m_ResourceSignatureCache.emplace(HashMapStringKey{pResourceSignature->GetDesc().Name, false}, pResourceSignature);
+
+                return pResourceSignature.Detach();
```

Comment:
Detaching here will result in memory leaks as the reference will be lost.
Detaching means transferring ownership to another object. Here the ownership is lost